### PR TITLE
Add mergify merging queues

### LINKED ISSUE #3201: Implement a merge queue for Zebra using the bors merge bot
## Motivation

We want to speed up CI and at the same time ensure that code changes to zebra are merged atomically, so that we know that our main branch always contains a version of the code that has been fully tested.

### Designs

We can do this by having a merge queue for Zebra, using the bors merge bot (see https://kflansburg.com/posts/merge-queues/ for a good intro) and:

1. Adding a staging branch
2. Splitting CI jobs between staging-only and "every PR change"
3. Only running macOS and Windows tests on the staging branch
## Related Work

See #1533 also

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation
1. Speed up merge times by prioritizing, queueing, and automatically merging Pull Requests 
2. Parallelize the CI process while in the queue, to reduce the time it takes for multiple PRs to be merged

## Solution

Previous:
- Use bors-ng was the proposed solution at first to solve this problem (#3201), but this required creating extra branches, using public bots, and some additional constraints which made Mergify a better option.

Actual
- Use Mergify as it respects Githubs branch protection rules, it also allows for multiple actions to help with the queueing process https://docs.mergify.com/actions/

This PR must ensures that the actual labels for prioritization are being used (5 in total) and distributed in 3 different queues (urgent, medium and default)
- :ambulance: P-Critical --> urgent
- :fire: P-High --> urgent
- :zap: P-Medium --> medium
- :snowflake: P-Low --> default
- :sparkles: P-Optional --> default

For a PR to be assigned to a queue all these conditions must be met:
- Targeting the **main** branch
- Having one or more reviews
- CI tests must be green
- Not being a `draft` PR
- Not having the labels ~~work-in-progress~~ or `do-not-merge`

## Review
@dconnolly @teor2345 

### Reviewer Checklist
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work
- Use our bot
- Consider dependabot on the configuration
- Add extra actions

Closes #3201

### CODE REVIEW

File: .github/mergify.yml
Comment (teor2345) [inline .github/mergify.yml:7]: Test Coverage: Can we also check the Google Cloud test result for each queue and rule?
(I think it's called `pull-request`.)

File: .github/mergify.yml
Comment (teor2345) [inline .github/mergify.yml:26]: Scope: It seems like we're missing the automated `merge` action.

I think that's a good idea while we are testing mergify, but I just wanted to mention it so other reviewers aren't confused.

We will also need to check the cached state tests in each queue before we auto-merge anything. But that can happen in a later PR.

File: .github/mergify.yml
Comment (teor2345) [inline .github/mergify.yml:30]: Surprising Behaviour: Should we also skip "do not merge" labels here?

File: .github/mergify.yml
Comment (teor2345) [inline .github/mergify.yml]: Streamlined Defaults: Can we update PRs by default, without requiring a specific label?
```suggestion
```

File: .github/mergify.yml
Comment (teor2345) [inline .github/mergify.yml]: Duplication: Can we replace the "work-in-progress" label with the "draft" PR status?

It seems redundant to have both.

**Review by teor2345 — COMMENTED** [review]
This looks pretty good.

I have some suggestions that will help Mergify match our current workflow a bit more closely. But I'm happy for you and @dconnolly to do whatever you think is best here.

File: .github/mergify.yml
Comment (gustavovalverde) [reply .github/mergify.yml:26]: The merge method is automatically considered when a PR is sent into a queue; the default is *merge*

Now, the only extra consideration to take is which Merge method to use. Possible values are **merge**, **squash** or **rebase**.

cc: @dconnolly

File: .github/mergify.yml
Comment (gustavovalverde) [reply .github/mergify.yml:30]: Done

**Comment by gustavovalverde** [comment]
I've gone through all the recommendations given by Teor. I'll be awaiting your review to merge afterward @dconnolly

**Review by dconnolly — APPROVED** [review]
Looks great! Excited to try this out

**Pull request merged**
